### PR TITLE
Add locking to SecureTransport P/Invokes

### LINKED ISSUE #24821: ClientWebSocket produces malloc exception in multi-threaded environments.
Running a .NET Core 2.0 CLI app on macOS 10.12.6. ClientWebSocket version is the latest I believe.

When operating in a multi-threaded environment, I receive what looks like a C++ malloc exception.

This appears to be happening when the remote host terminates the connection unexpectedly. I have locks on reads + writes so there is never more than one of each happening at once. When I make a call to either `SendAsync` or `ReceiveAsync` and the connection has been terminated unexpectedly, I receive the following error:

```
dotnet(75902,0x700010355000) malloc: *** error for object 0x7fe5669dbf20: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
```

I can't say for certain whether or not the remote host has terminated the connection because I don't have a stack trace, but it is my suspicion as I know there is a rate limit on the web socket endpoint, and this issue only happens shortly after sustained requests exceed rate limit threshold.

I can't think of any way of preventing this issue in user code (other than a rate limiter but that only stops a certain type of unexpected termination.)  Adding a lock won't solve it because the the connection could be terminated between when I call ReceiveAsync and when ReceiveAsync reaches the point in native code where it breaks.

### PR DESCRIPTION
**Pull request by bartonjs**
Makes the macOS version of SslStream capable of doing one read and one write "at the same time" without taking down the process.

Fixes #24821 (and perhaps others).
Best tested by #24759 (TBD)

### CODE REVIEW

**Comment by wfurt** [comment]
looks ok to me. The lock is per stream, right?

**Comment by bartonjs** [comment]
>The lock is per stream, right?

Yep.  Each different SslStream can be used for one P/Invoke at a time.

**Comment by stephentoub** [comment]
Thanks, @bartonjs.  Generally looks good.
- I seem to remember there being some code somewhere that dup'd crypto handles.  Do we have to worry about the possibility that there may be two different SafeHandles (presumably with ownsHandle:false) for the same underlying native ptr?  Or maybe a pointer that represents some sub state of a larger object, each of which would be represented by two different handles?  I know somewhere there was code for parent/child SafeHandles, but maybe that was just in the OpenSSL implementation?
- Any concerns here about deadlock?  Presumably not, but I'm just wondering if there might be a situation where, for example, the native lib makes a reverse P/Invoke call to something that schedules work and then blocks waiting for it.

File: src/System.Net.Security/src/System/Net/Security/SslStreamPal.OSX.cs [hunk 1]
```diff
@@ -259,7 +273,13 @@ public static void VerifyPackageInfo()
                     sslContext.Write(inputBuffer.token, inputBuffer.offset, inputBuffer.size);
                 }
 
-                SecurityStatusPal status = PerformHandshake(sslContext.SslContext);
+                SecurityStatusPal status;
+                SafeSslHandle sslHandle = sslContext.SslContext;
```
Comment (stephentoub) [inline src/System.Net.Security/src/System/Net/Security/SslStreamPal.OSX.cs]: Nit: swap these two lines?

**Comment by stephentoub** [comment]
ps Once this is in, we should port it to release/2.0.0.

**Comment by bartonjs** [comment]
> I seem to remember there being some code somewhere that dup'd crypto handles. ... maybe that was just in the OpenSSL implementation?

Yeah, I think just the OpenSSL implementation.  As far as I remember / understand, the SslHandle object is 1:1 with the SslStream (except in degenerate 0:1 cases where SslStream was never really initialized

> Any concerns here about deadlock? ... reverse P/Invoke ...

The SSLRead and SSLWrite functions can call callbacks, they're implemented as enqueuing or dequeuing bytes to a ready state.  But there's nothing where we call them, they call us, and we call them again as part of the same operation.

Both valid concerns, thankfully neither applies.

**Comment by bartonjs** [comment]
@dotnet-bot Test Linux x64 Release Build please

**Comment by bartonjs** [comment]
@dotnet-bot Test Linux x64 Release Build please

```
Unhandled Exception: System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at System.Buffers.TlsOverPerCoreLockedStacksArrayPool`1.Return(T[] array, Boolean clearArray) in /root/coreclr/src/mscorlib/shared/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs:line 150
   at System.Net.Security.SslStreamInternal.Finalize()
```
https://mc.dot.net/#/user/bartonjs/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/8dfded4df187ee2933a4e177aa1891e02f1cbf75/workItem/System.Net.Security.Tests/wilogs

(fedora.25)

**Comment by stephentoub** [comment]
I expect that failure is another manifestation of https://github.com/dotnet/corefx/issues/24775

**Pull request merged**